r2-core/src/main/java/com/linkedin/r2/message/BaseRequest.java

```diff
@@ -18,11 +18,24 @@
 
   protected BaseRequest(Map<String, String> headers, List<String> cookies, URI uri, String method)
   {
-    super(headers, cookies);
-    ArgumentUtil.notNull(uri, "uri");
-    ArgumentUtil.notNull(method, "method");
-    _uri = uri;
-    _method = method;
+    this(headers, cookies, uri, method, true);
+  }
+
+  protected BaseRequest(Map<String, String> headers, List<String> cookies, URI uri, String method, boolean validateArgs)
+  {
+    super(headers, cookies, validateArgs);
+    if (validateArgs)
+    {
+      ArgumentUtil.notNull(uri, "uri");
```

Comment:
Is this really costly? Not sure if it is worth to use flag to control such minor CPU.

Reply:
Refer to previous comment about why it's expensive

Reply:
In this particular routine, you are just saving two Null checks? I didn't see them stand out in your flamegraph, or maybe I overlooked something in the flamegraph.

Reply:
I don't think we care about map sorting at all! What we care about is case-insensitivity to follow the http header spec. TreeMap just happens to be a sorted map that can be configured as a case -insensitive map.

The entire point of validateArgs is this: We will not validate/copy anything and trust whatever the subclass sent us if this is false. This is supposed to be carefully used by subclasses who know what they are doing (like us creating a url rewritten request copy here).

The default value of validateArgs is true, and in this mode all the existing validation and copying checks run.

Reply:
Most of the use here is to proxy the validateArgs to the superclass (BaseMessage) where all the expensive header copy and field name validation happens. Thought while we are it might as well also save on the null checks.